File: A1/packages/baumwelch/BaumWelch.java
package baumwelch;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import matrix.Matrix;
import hmmtoolbox.Toolbox;

public class BaumWelch {
    private Matrix A;
    private Matrix B;
    private Matrix pi;
    private Matrix emiSeq;

    public BaumWelch(Matrix A, Matrix B, Matrix pi, Matrix emissionSequence) {
        this.A = A;
        this.B = B;
        this.pi = pi;
        this.emiSeq = emissionSequence;
    }

    public Hashtable<String, Matrix> baumWelch(List<String> input) {
        int T = emiSeq.m;
        int N = A.m;
        int M = B.n;
        int maxIters = 10000;
        double oldLogProb = Double.MIN_VALUE;

        double tol = 0.01;
        int z = 0;

        for (z = 0; z < maxIters; z++) {

            List<Matrix> alphas = new ArrayList<>();
            List<Double> scales = new ArrayList<>();

            double c0 = 0;

            Matrix a0 = new Matrix(N, 1);
            for (int i = 0; i < N; i++) {
                a0.set(i, 0, pi.get(0, i) * B.get(i, (int) emiSeq.get(0, 0)));
                c0 = c0 + a0.get(i, 0);
            }

            c0 = 1 / c0;
            a0.scalarMult(c0);

            scales.add(c0);
            alphas.add(a0);

            for (int t = 1; t < T; t++) {
                double ct = 0;

                Matrix at = new Matrix(N, 1);
                for (int i = 0; i < N; i++) {
                    double ati = 0;
                    for (int j = 0; j < N; j++) {
                        ati = ati + alphas.get(t - 1).get(j, 0) * A.get(j, i);
                    }
                    ati = ati * B.get(i, (int) emiSeq.get(t, 0));
                    at.set(i, 0, ati);
                    ct += at.get(i, 0);
                }

                ct = 1 / ct;
                at = at.scalarMult(ct);
                alphas.add(at);
                scales.add(ct);
            }

            List<Matrix> betas = new ArrayList<>();

            Matrix BTm1 = new Matrix(N, 1);
            for (int i = 0; i < N; i++)
                BTm1.set(i, 0, scales.get(scales.size() - 1));

            betas.add(BTm1);

            for (int t = T - 2; t >= 0; t--) {

                Matrix bt = new Matrix(N, 1);

                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < N; j++) {
                        bt.set(i, 0, bt.get(i, 0)
                                + A.get(i, j) * B.get(j, (int) emiSeq.get(t + 1, 0)) * betas.get(0).get(j, 0));
                    }
                }

                bt = bt.scalarMult(scales.get(t));
                betas.add(0, bt);
            }

            List<Matrix> digammas = new ArrayList<>();
            List<Matrix> gammas = new ArrayList<>();

            for (int t = 0; t < T - 1; t++) {
                double denom = 0;
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < N; j++) {
                        denom += alphas.get(t).get(i, 0) * A.get(i, j) * B.get(j, (int) emiSeq.get(t + 1, 0))
                                * betas.get(t + 1).get(j, 0);
                    }
                }
                Matrix digamma = new Matrix(N, N);
                Matrix gamma = new Matrix(N, 1);
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < N; j++) {
                        digamma.set(i, j, alphas.get(t).get(i, 0) * A.get(i, j) * B.get(j, (int) emiSeq.get(t + 1, 0))
                                * betas.get(t + 1).get(j, 0) / denom);
                        gamma.set(i, 0, gamma.get(i, 0) + digamma.get(i, j));
                    }
                }

                digammas.add(digamma);
                gammas.add(gamma);
            }

            double denom = 0;
            for (int i = 0; i < N; i++) {
                denom += alphas.get(T - 1).get(i, 0);
            }

            Matrix gamTm1 = new Matrix(N, 1);
            for (int i = 0; i < N; i++) {
                gamTm1.set(i, 0, alphas.get(T - 1).get(i, 0) / denom);
            }

            gammas.add(gamTm1);

            for (int i = 0; i < N; i++) {
                pi.set(0, i, gammas.get(0).get(i, 0));
            }

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    double numer = 0;
                    double denom2 = 0;
                    for (int t = 0; t < T - 1; t++) {
                        numer = numer + digammas.get(t).get(i, j);
                        denom2 = denom2 + gammas.get(t).get(i, 0);
                    }
                    A.set(i, j, numer / denom2);
                }
            }

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    double numer = 0;
                    double denom2 = 0;
                    for (int t = 0; t < T; t++) {
                        if (emiSeq.get(t, 0) == j) {
                            numer = numer + gammas.get(t).get(i, 0);
                        }
                        denom2 = denom2 + gammas.get(t).get(i, 0);
                    }
                    B.set(i, j, numer / denom2);
                }
            }

            double logProb = 0;
            for (int t = 0; t < T; t++) {
                logProb = logProb + Math.log10(scales.get(t));
            }
            logProb = -logProb;

            if (Math.abs(logProb - oldLogProb) < tol) {
                break;
            }
            oldLogProb = logProb;
            // System.out.println(oldLogProb);
        }

        System.out.println("Iterations: " + z);

        Hashtable<String, Matrix> retMap = new Hashtable<String, Matrix>();
        retMap.put("A", A);
        retMap.put("B", B);
        retMap.put("pi", pi);

        return retMap;
    }
}

File: A1/include/baumwelch.cpp
#include "baumwelch.h"

BaumWelch::BaumWelch(Matrix A, Matrix B, Matrix pi, Matrix emissionSequence) 
{
    this->A = A;
    this->B = B;
    this->pi = pi;
    this->emissionSequence = emissionSequence;
}

std::vector<Matrix> BaumWelch::baumwelch()
{
    int T = emissionSequence.getRowLength();
    int N = A.getRowLength();
    int M = B.getColumnLength();
    int maxIters = 100;
    double oldLogProb = -std::numeric_limits<double>::max();

    double tolerance = 0.01;
    int z = 0;

    for (z = 0; z < maxIters; z++)
    {
        std::vector<Matrix> alphas;
        alphas.reserve(T);
        std::vector<double> scales;
        scales.reserve(T);

        double c0 = 0;

        Matrix a0(N, 1);
        for (int i = 0; i < N; i++)
        {
            a0.set(i, 0, pi.get(0, 1) * B.get(i, getEmission(0)));
            c0 = c0 + a0.get(i, 0);
        }

        c0 = 1.0 / c0;
        a0.scalarMult(c0);

        scales.push_back(c0);
        alphas.push_back(a0);

        for (int t = 1; t < T; t++)
        {
            double ct = 0.0;

            Matrix at(N, 1);
            for (int i = 0; i < N; i++)
            {
                double ati = 0.0;
                for (int j = 0; j < N; j++)
                {
                    ati += alphas[t - 1].get(j, 0) * A.get(j, i);
                }
                ati *= B.get(i, getEmission(t));
                at.set(i, 0, ati);
                ct += at.get(i, 0);
            }

            ct = 1.0 / ct;
            at = at.scalarMult(ct);
            alphas.push_back(at);
            scales.push_back(ct);
        }

        std::vector<Matrix> betas;
        betas.reserve(T);

        Matrix BTm1(N, 1);
        for (int i = 0; i < N; i++)
        {
            BTm1.set(i, 0, scales[(scales.size() - 1)]);
        }

        betas.push_back(BTm1);

        for (int t = T - 2; t >= 0; t--)
        {
            Matrix bt(N, 1);

            for (int i = 0; i < N; i++)
            {
                for (int j = 0; j < N; j++)
                {
                    bt.set(i, 0, bt.get(i, 0) + A.get(i, j) * B.get(j, getEmission(t + 1)) * betas[0].get(j, 0));
                }
            }

            bt = bt.scalarMult(scales[t]);
            std::vector<Matrix> newBetas;
            newBetas.push_back(bt);
            newBetas.insert(std::end(newBetas), std::begin(betas), std::end(betas));
            betas = newBetas;
            // betas.push_back(bt);
            // betas.insert(betas.begin(), bt);
        }

        std::vector<Matrix> digammas;
        digammas.reserve(T);
        std::vector<Matrix> gammas;
        gammas.reserve(T);

        for (int t = 0; t < T - 1; t++)
        {
            double denom = 0;
            for (int i = 0; i < N; i++)
            {
                for (int j = 0; j < N; j++)
                {
                    denom += alphas[t].get(i, 0) * A.get(i, j) * B.get(j, getEmission(t + 1)) * betas[t + 1].get(j, 0);
                }
            }
            Matrix digamma(N, N);
            Matrix gamma(N, 1);
            for (int i = 0; i < N; i++)
            {
                for (int j = 0; j < N; j++)
                {
                    digamma.set(i, j, alphas[t].get(i, 0) * A.get(i, j) * B.get(j, getEmission(t + 1)) * betas[t + 1].get(j, 0) / denom);
                    gamma.set(i, 0, gamma.get(i, 0) + digamma.get(i, j));
                }
            }

            digammas.push_back(digamma);
            gammas.push_back(gamma);
        }

        double denom = 0;
        for (int i = 0; i < N; i++)
        {
            denom += alphas[T - 1].get(i, 0);
        }

        Matrix gamTm1(N, 1);
        for (int i = 0; i < N; i++)
        {
            gamTm1.set(i, 0, alphas[T - 1].get(i, 0) / denom);
        }

        gammas.push_back(gamTm1);

        for (int i = 0; i < N; i++)
        {
            pi.set(0, i, gammas[0].get(i, 0));
        }

        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < N; j++)
            {
                double numer = 0.0;
                double denom2 = 0.0;
                for (int t = 0; t < T - 1; t++)
                {
                    numer = numer + digammas[t].get(i, j);
                    denom2 = denom2 + gammas[t].get(i, 0);
                }
                A.set(i, j, numer / denom2);
            }
        }

        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < M; j++)
            {
                double numer = 0;
                double denom2 = 0;
                for (int t = 0; t < T; t++)
                {
                    if (getEmission(t) == j)
                    {
                        numer = numer + gammas[t].get(i, 0);
                    }
                    denom2 = denom2 + gammas[t].get(i, 0);
                }
                B.set(i, j, numer / denom2);
            }
        }

        double logProb = 0.0;
        for (int t = 0; t < T; t++)
        {
            logProb = logProb + std::log10(scales[t]);
        }
        logProb = -logProb;

        // std::cout << std::fabs(logProb - oldLogProb) << std::endl;
        // if (std::fabs(logProb - oldLogProb) < tolerance)
        // {
        //     break;
        // }
        oldLogProb = logProb;
        std::cout << z << std::endl;
    }

    std::cout << "Iterations: " << z << std::endl;

    std::vector<Matrix> retVec;

    retVec.push_back(A);
    retVec.push_back(B);
    retVec.push_back(pi);

    return retVec;
}

int BaumWelch::getEmission(int t) {
    if (t > emissionSequence.getRowLength())
        throw new std::logic_error("t not matching emission sequence length!");

    return (int) emissionSequence.get(t, 0);
}